### ISSUE TITLE
Chore: External table test flake with flight sql

### ISSUE THREAD

**Issue by scsmithr** [GlareDB/glaredb#2366]
## Description

```
02:06:01.333 ERROR testing::slt::cli: Error while running test `sqllogictests/external_table` error=test fail: query result mismatch:
[SQL] select count(*) from glare_catalog.tables where builtin = false;
[Diff] (-expected|+actual)
-   1
+   4
at /home/runner/work/glaredb/glaredb/testdata/sqllogictests/external_table.slt:30
```

**Comment by universalmind303** [comment]
@scsmithr I just tried to reproduce this on commit `66c2f6bc` (main) and was not able to reproduce this 


```
> just slt 'sqllogictests/external_table' --protocol=flightsql  

cargo build --bin glaredb 
    Finished dev [unoptimized + debuginfo] target(s) in 0.70s
just slt-bin sqllogictests/external_table --protocol=flightsql
./target/debug/glaredb sql-logic-tests sqllogictests/external_table --protocol=flightsql
15:13:05.648  INFO metastore::local: starting local metastore path="/var/folders/tt/gj5kpg1d6q76s1m4k1s3xjlw0000gn/T/.tmpjQiogY"
15:13:05.650  INFO metastore::srv: Creating new Metastore service process_id=9b85211b-d974-406a-badb-77598449e950
15:13:05.656  INFO glaredb::server: Starting GlareDB 0.8.0
Connect via RPC: grpc://0.0.0.0:49660

15:13:05.656  INFO glaredb::server: enabling flight sql service
15:13:05.657  INFO glaredb::args::slt: Running test: `sqllogictests/external_table`
Tests took 145.758375ms to run
```

was this only observed in CI?

**Comment by tychoish** [comment]
How many times did you run the test?

**Comment by tychoish** [comment]
I think we should: 
- run this 1000 times and see if we can provoke a failure. Locally or in CI (if we can just run a tight loop around this) 
- do a bit a poking around the glaredb catalog -> flightsql code path. I think the interesting thing here is that in sean's initial copy of the test failure, the catalog has _more_ tables than it should, which makes me think that there's just some stale data that's hanging around.

Potential theories, lacking any other evidence:
- there's something stateful in the test harness: a previous test created some external tables that aren't being cleaned up, or that expire when run against other harnesses.
- the glare catalogue could be cached in the flightsql service or session, so we have a stale view.
- for completeness: "something about the query just executes incorrectly on flightsql," and that seems much less likely, but if we're doing any kind of special handling or parsing of queries in this code path then maybe that would be a reasonable explanation. 

I think if we can't get it to reproduce in 1000 runs, and a couple of us read through the code for the above three options, then it's probably worth closing this and calling it good for now. It could resurface, but I'm less worried because the failure is "there are more things in the catalog than we expect" and not _fewer things that we expect._ both are potentially bad, but the former might not represent data loss, where the later definitely is. Particularly because most of the explanations that I can come up with are basically "there's a problem with the test" rather than "there's a problem with the implementation" (or at least 2/3rds are) and it's extremely rare... that's something at least.

**Comment by scsmithr** [comment]
A separate flake: https://github.com/GlareDB/glaredb/actions/runs/7672628076/job/20913590679?pr=2498

These failures lead me to believe there's something in the flightsql code path that's intermingling databases when they shouldn't be.

**Comment by tychoish** [comment]
https://github.com/glaredb/glaredb/blob/main/crates/rpcsrv/src/flight/handler.rs#L156 <- this is the obvious suspicious line. I've read through basically all of the flight code that I could find, and nothing else jumped out even a little.

**Comment by vrongmeal** [comment]
> https://github.com/glaredb/glaredb/blob/main/crates/rpcsrv/src/flight/handler.rs#L156 <- this is the obvious suspicious line. I've read through basically all of the flight code that I could find, and nothing else jumped out even a little.
> 
> the proxy bits seem like they would error if a db_name wasn't present, but if you connect directly (as I suspect these tests do), seems possible to trip this.

Yup! This is something I was trying to find during my initial investigation. Will look more into it. There's a good chance this is it.
